model/src/model.py

```diff
@@ -80,19 +121,130 @@ def forward(self, data: Optional[torch.Tensor] = None) -> tuple[torch.Tensor, to
             current = self.forward_weights(data)
 
         spk, mem = self.forward_lif(current, self.mem)
-        self.spk_rec.append(spk)
+        self.mem = mem
         self.mem_rec.append(mem)
 
+        logging.debug("")
+        logging.debug(f"current: {str(current)}")
+        logging.debug(f"mem: {str(mem)}")
+        logging.debug(f"spk: {str(spk)}")
+
         return spk, mem
 
     def train_forward(self, data: Optional[torch.Tensor] = None) -> None:
-        self.optimizer.zero_grad()
-
-        spk, _mem = self.forward(data)
-        loss = torch.sum(spk)
-        loss.backward()
-
-        self.optimizer.step()
+        """
+        The LPL learning rule is implemented here. It is defined as dw_ji/dt,
+        for which we optimize the computation with matrices.
+
+        This learning rule is broken up into three terms:
+
+         1. The first term contains S_j(t) * f'(U_i(t)). We form a matrix via
+            some unsqueezes to form a first term matrix of size (batch_size, j,
+            i). This is mainly done by matrix multiplication.
+
+         2. The second term contains S_i. We form a matrix via some unsqueezes
+            to form a second term matrix of size (batch_size, i, j). This is
```

Comment:
confirming that this should be a row duplication procedure.

Reply:
yep that is correct